File: UI/LatestNGApp/PMSUIApp/src/app/Services/project.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable, of } from 'rxjs';
import { Projectlist } from '../Models/projectlist';
import { HttpClient } from '@angular/common/http';
import { Project } from '../Models/project';
import { Projecttask } from '../Models/projecttask';
import { Projectname } from '../Models/projectname';
import { Addproject } from '../Models/addproject';
import { Unassignedtask } from '../Models/unassignedtask';
import { Assignedtaskbymanager } from '../Models/assignedtaskbymanager';

@Injectable({
  providedIn: 'root'
})
export class ProjectService {


  constructor(private httpClient: HttpClient) { }
  private selectedProjectIdSubject = new BehaviorSubject<any>(null);
  selectedProjectId$ = this.selectedProjectIdSubject.asObservable();

  setSelectedProjectId(id: number | null) {
    this.selectedProjectIdSubject.next(id)
  }

  getProjectsList(teamMemberId: number): Observable<Projectlist[]> {
    let url = "http://localhost:5248/api/projects/list/" + teamMemberId
    return this.httpClient.get<Projectlist[]>(url)
  }

  getProjectDetails(projectId: number): Observable<Project> {
    let url = "http://localhost:5248/api/projects/" + projectId
    return this.httpClient.get<Project>(url)
  }

  getProjectMembers(projectId: number): Observable<number[]> {
    let url = "http://localhost:5248/api/projects/teammembers/" + projectId
    return this.httpClient.get<number[]>(url)
  }

  getTasksOfProject(projectId: number, timePeriod: string): Observable<Projecttask[]> {
    let url = "http://localhost:5248/api/projects/tasks/" + projectId + "/" + timePeriod
    return this.httpClient.get<Projecttask[]>(url)
  }
  getProjectNames(employeeId: number): Observable<Projectname[]> {
    let url = "http://localhost:5248/api/projects/employee/" + employeeId
    return this.httpClient.get<Projectname[]>(url)
  }

  getManagerProjects(managerId: number): Observable<Projectlist[]> {
    let url = "http://localhost:5248/api/projects/manager/" + managerId
    return this.httpClient.get<Projectlist[]>(url)
  }

  addProject(project: Addproject): Observable<boolean> {
    let url = "http://localhost:5248/api/projects/addproject"
    return this.httpClient.post<boolean>(url, project)
  }
  updateProject(project: Project): Observable<boolean> {
    let url = "http://localhost:5248/api/projects"
    return this.httpClient.put<boolean>(url, project)
  }
  deleteProject(projectId: number): Observable<boolean> {
    let url = "http://localhost:5248/api/projects/" + projectId
    return this.httpClient.delete<boolean>(url)
  }     

  unAssignedTask(projectId: number, timePeriod: string): Observable<Unassignedtask[]> {
    let url = "http://localhost:5248/api/projects/unassignedtask/" + projectId + "/" + timePeriod
    return this.httpClient.get<Unassignedtask[]>(url)
  }
  assignedTasksByManager(managerId:number,timePeriod:string):Observable<Assignedtaskbymanager[]>{
    let url="http://localhost:5248/api/projects/assignedtasksbymanager/" + managerId + "/" +timePeriod
    return this.httpClient.get<Assignedtaskbymanager[]>(url)
  }
}

File: API/Projects/Services/ProjectService.cs
using Transflower.PMSApp.Projects.Entities;
using Transflower.PMSApp.Projects.Models;
using Transflower.PMSApp.Projects.Repositories.Interfaces;
using Transflower.PMSApp.Projects.Services.Interfaces;

namespace Transflower.PMSApp.Projects.Services;
public class ProjectService : IProjectService
{
    private readonly IProjectRepository _repository;

    public ProjectService(IProjectRepository repository)
    {
        _repository = repository;
    }
    public async Task<List<Project>> GetAll()
    {
        return await _repository.GetAll();
    }

    public async Task<Project> GetById(int projectId)
    {
        return await _repository.GetById(projectId);
    }
    public async Task<List<ProjectList>> GetProjectsList(int teamMemberId){
        return await _repository.GetProjectsList(teamMemberId);
    }

    public  async Task<bool> Insert(Project project)
    {
        return await _repository.Insert(project);
    }

    public async Task<bool> Update(Project project)
    {
        return await _repository.Update(project);
    }
        public async Task<bool> Delete(int projectId)
    {
        return await _repository.Delete(projectId);
    }

    public async Task<List<int>> GetProjectMembers(int projectId)
    {
        return await _repository.GetProjectMembers(projectId);
    }

    public async Task<List<ProjectTask>> GetTasksOfProject(int projectId,string timePeriod)
    {
        return await _repository.GetTasksOfProject(projectId,timePeriod);
    }

    public async Task<List<ProjectName>> GetProjectNames(int employeeId)
    {
        return await _repository.GetProjectNames(employeeId);
    }

    public async Task<List<ProjectList>> GetManagerProjects(int managerId)
    {
        return await _repository.GetManagerProjects(managerId);
    }
    public async Task<List<UnAssignedTask>> GetUnAssignedTasks(int projectId, string timePeriod)
    {
        return await _repository.GetUnAssignedTasks(projectId,timePeriod);
    }

    public async Task<List<AssignedTaskByManager>> GetAssignedTasksByManager(int managerId, string timePeriod)
    {
        return await _repository.GetAssignedTasksByManager(managerId,timePeriod);
    }







}

